internal/pkg/gateway/api.go

```diff
@@ -134,26 +134,42 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
 	}
 	proposal, err := protoutil.UnmarshalProposal(signedProposal.ProposalBytes)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "failed to unpack transaction proposal: %s", err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-	channel, chaincodeID, hasTransientData, err := getChannelAndChaincodeFromSignedProposal(signedProposal)
```

Comment:
so you remove the call to `getChannelAndChaincodeFromSignedProposal` and now you have the header, chaincode proposal payload at hand, to pass into `prepareTransaction`. 

So instead of doing unmarshaling of the proposal at CreateSignedTx and here, you just do it once.
This addresses (2) and I can understand why this is more efficien. 

I don't understand why:
> Instead of collecting and holding copies of all proposal responses in order to reuse the protoutil.CreateTx() function, just store a single response payload and all of the signed endorsement objects. 

Is worth it. Once you obtain the *ProposalResponse you have a pointer which you then pass "by reference" to a function. You don't copy the memory of the payload, as it is []byte and it's also passed "by reference" (you pass a pointer to an array when you do the function call, and not copy the array itself).
Where do you get the efficiency improvement from?

Reply:
I understand it's passed by reference.  The idea of this change was to let go of any references to structures that are no longer needed (and hence available for GC) while it potentially continues to collect more endorsements.  This might allow more concurrent clients to make calls to Endorse() before running out of memory.

Reply:
OK so you're saying that the memory corresponding to the payload can be GCed early, I see.